CK Tile GEMM Compute V2 (2 LDS Ping Pong mechanism)

ThomasNing opened the pull request:
## Proposed changes

Enables the CK Tile GEMM Compute Version 2 pipeline. It has the same algorithm as the Old CK pipeline V4. It has performance improvement comparing with Compute V1 when they have the same block tile. This PR also has some minor changes in changing the block tile of Compute Version 1 to improve the performance and add an initialization method of the GEMM input.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

ThomasNing committed:
comp v4 setup

ThomasNing committed:
add a file

ThomasNing committed:
Finished the coding of the feature, Compiler not in the way we supposed to have

ThomasNing committed:
Update some of the code to better format

ThomasNing committed:
get tback the restrict variable name, need to switch out to solve the transpose issue

ThomasNing committed:
Solve the compiler issue on SHMEM conflict

ThomasNing committed:
Finish the feature and merge with develop on the computeV2

ThomasNing committed:
roll back to compute pipeline

ThomasNing committed:
Add the changes from include/ck_tile

aosewski commented on example/ck_tile/03_gemm/universal_gemm.cpp:
```diff
@@ -196,6 +217,15 @@ float gemm_calc(const ck_tile::GemmHostArgs& args, const ck_tile::stream_config&
+                Run(ck_tile::bool_constant<true>{},
+                    ck_tile::integral_constant<ck_tile::TailNumber, ck_tile::TailNumber::Two>{});
+            }
+        }
```
Is there any `else` logic? Maybe we should throw exception if other configuration is not supported?

ThomasNing replied:
Solved.

aosewski commented on example/ck_tile/03_gemm/gemm_basic.hpp:
```diff
@@ -13,6 +13,7 @@
 
 #define CK_TILE_PIPELINE_COMPUTE 1
 #define CK_TILE_PIPELINE_MEMORY 2
+#define CK_TILE_PIPELINE_COMPUTE_V2 3
```
Let's align those names with actuall pipeline name. 
```suggestion
#define CK_TILE_PIPELINE_COMPUTE_V4 3
```

ThomasNing replied:
Solved.

aosewski commented on example/ck_tile/03_gemm/gemm_basic.hpp:
```diff
@@ -13,6 +13,7 @@
 
 #define CK_TILE_PIPELINE_COMPUTE 1
```
Let's align those names with actuall pipeline name.
```suggestion
#define CK_TILE_PIPELINE_COMPUTE_V3 1
```

aosewski commented on example/ck_tile/03_gemm/gemm_basic.hpp:
```diff
@@ -22,10 +23,17 @@
 #define GEMM_PIPELINE ck_tile::GemmPipelineAgBgCrMem
 #define UNIVERSAL_GEMM_PIPELINE ck_tile::BaseGemmPipelineAgBgCrMem
 #define GEMM_PIPELINE_SCHEDULER ck_tile::GemmPipelineScheduler::Interwave
+constexpr bool isDoubleSmemBuffer = false;
```
Would align better with other parameters naming. 
```suggestion
constexpr bool DoubleSmemBuffer = false;
```

aosewski commented on include/ck_tile/ops/gemm/pipeline/tile_gemm_traits.hpp:
```diff
@@ -10,6 +10,7 @@ namespace ck_tile {
 template <bool kPadM_,
           bool kPadN_,
           bool kPadK_,
+          bool isDoubleSmemBuffer_,
```
Is it used on other pipelines than `universal` ones? If not then this is not necessary here.

aosewski commented on include/ck_tile/ops/gemm/pipeline/tile_gemm_traits.hpp:
```diff
@@ -32,6 +35,7 @@ struct TileGemmTraits
 template <bool kPadM_,
           bool kPadN_,
           bool kPadK_,
+          bool isDoubleSmemBuffer_,
```
```suggestion
          bool DoubleSmemBuffer_,
```

aosewski commented on include/ck_tile/ops/gemm/pipeline/tile_gemm_traits.hpp:
```diff
@@ -42,6 +46,8 @@ struct TileGemmUniversalTraits
     static constexpr bool kPadN = kPadN_;
     static constexpr bool kPadK = kPadK_;
 
+    static constexpr bool isDoubleSmemBuffer = isDoubleSmemBuffer_;
```
```suggestion
    static constexpr bool DoubleSmemBuffer = isDoubleSmemBuffer_;
```

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -474,14 +474,14 @@ struct GemmKernel
-                                       const SplitKBatchOffset& splitk_batch_offset,
-                                       const index_t block_idx_m,
-                                       const index_t block_idx_n)
+    CK_TILE_DEVICE static void RunGemmSinglePointer(const ADataType* a_ptr,
```
This name is really confusing. My first thought was that there'd be multiple tensors pointers for each input... 
```suggestion
    CK_TILE_DEVICE static void RunGemm(const ADataType* a_ptr,
```

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -474,14 +474,14 @@ struct GemmKernel
+    CK_TILE_DEVICE static void RunGemmSinglePointer(const ADataType* a_ptr,
+                                                    const BDataType* b_ptr,
+                                                    CDataType* c_ptr,
+                                                    void* smem_ptr_0,
+                                                    const GemmKernelArgs& kargs,
+                                                    const SplitKBatchOffset& splitk_batch_offset,
```
Please update doc.

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -495,8 +495,64 @@ struct GemmKernel
+        }
+    }
+
+    /**
+     * @brief Runs single GEMM problem cooperatively by whole workgroup.
+     *
+     * @param a_ptr input A pointer
+     * @param b_ptr input B pointer
+     * @param c_ptr output C pointer
+     * @param kargs GEMM kernel arguments
+     * @param block_idx_m The GEMM's output M dimension tile index processed by this workgroup.
+     * @param block_idx_n The GEMM's output N dimension tile index processed by this workgroup.
+     *
+     * @tparam DstInMemOp Destination memory operation (default: set).
+     */
```
Please update doc.

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -495,8 +495,64 @@ struct GemmKernel
+        // Run Epilogue Pipeline
+        auto& c_block_window = gemm_tile_windows.at(I2);
+
+        constexpr bool is_output_c_reg_transposed =
+            EpiloguePipeline::IsOutputTransposed() != GemmPipeline::IsTransposeC();
+        if constexpr((DstInMemOp == memory_operation_enum::set) || (sizeof(CDataType) > 2) ||
+                     (GemmPipeline::VectorSizeC % 2 == 0 &&
+                      std::is_same_v<CLayout, tensor_layout::gemm::RowMajor> &&
+                      is_output_c_reg_transposed))
+        {
+            EpiloguePipeline{}
+                .template operator()<decltype(c_block_window), decltype(c_block_tile), DstInMemOp>(
+                    c_block_window, c_block_tile);
+        }
```
Please merge with develop. This code was changed.

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -495,8 +495,64 @@ struct GemmKernel
+     * @tparam DstInMemOp Destination memory operation (default: set).
+     */
+    template <memory_operation_enum DstInMemOp = memory_operation_enum::set>
+    CK_TILE_DEVICE static void RunGemmDoublePointer(const ADataType* a_ptr,
```
```suggestion
    CK_TILE_DEVICE static void RunGemm2LDS(const ADataType* a_ptr,
```

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v4.hpp:
```diff
@@ -0,0 +1,551 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
```
The year convention is as follow: \<year-of-creation\>[-\<year-of-last-modification-if-different-from-year-of-creation\>]
```suggestion
// Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v4.hpp:
```diff
@@ -0,0 +1,551 @@
+};
+
+template <typename Problem, typename Policy = GemmPipelineAGmemBGmemCregComputeV4DefaultPolicy>
+struct GemmPipelineAgBgCrCompV4 : public BaseGemmPipelineAgBgCrCompV4<Problem>
```
It would be nice to have some doc here explaining what differentiates this pipeline form other pipelines as well as when this pipeline might be the best choice.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v4.hpp:
```diff
@@ -0,0 +1,551 @@
+                                                    ? B_LDS_Read_Inst_Num
+                                                    : B_LDS_Read_Inst_Num / 2;
+
+            constexpr auto num_ds_read_inst = num_ds_read_inst_a + num_ds_read_inst_b;
+
+            constexpr auto num_ds_write_inst = A_LDS_Write_Inst_Num + B_LDS_Write_Inst_Num;
+
+            constexpr auto num_buffer_load_inst = A_Buffer_Load_Inst_Num + B_Buffer_Load_Inst_Num;
+
+            constexpr auto num_issue = num_buffer_load_inst;
```
Please lets not make the code inflated with too much empty lines.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v4.hpp:
```diff
@@ -0,0 +1,551 @@
+                std::is_same_v<ADataType, remove_cvref_t<typename ADramBlockWindowTmp::DataType>> &&
+                    std::is_same_v<BDataType,
+                                   remove_cvref_t<typename BDramBlockWindowTmp::DataType>>,
+                "wrong!");
```
Please make error messages more verbose. At some point it maybe you who would get it :)

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,390 @@
+#include "ck_tile/ops/common/tensor_layout.hpp"
+
+namespace ck_tile {
+// Default policy for GemmPipelineAGmemBGmemCRegV1
+// Default policy class should not be templated, put template on member functions instead
```
Doc

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,390 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
```
Copyright year

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,390 @@
+namespace ck_tile {
+// Default policy for GemmPipelineAGmemBGmemCRegV1
+// Default policy class should not be templated, put template on member functions instead
+struct GemmPipelineAGmemBGmemCregComputeV4DefaultPolicy
```
What about deriving from `UniversalGemmPipelineAgBgCrPolicy` ?
https://github.com/ROCm/composable_kernel/blob/b2c7d7747e93dbb5c613aa8880f9fc942dd5e44a/include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp#L13
Most of the code is the same and you could hide base class functions with new implementation where needed.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,390 @@
+        constexpr index_t kKPerBlock = Problem::BlockGemmShape::kK;
+
+        // TODO: this 8 is AK1! should be a policy parameter!
+        constexpr auto a_lds_block_desc_0 = make_naive_tensor_descriptor(
+            make_tuple(number<kKPerBlock / 8>{}, number<kMPerBlock>{}, number<8>{}),
+            make_tuple(number<kMPerBlock * 8>{}, number<8>{}, number<1>{}),
+            number<8>{},
+            number<1>{});
+
+        constexpr auto a_lds_block_desc = transform_tensor_descriptor(
+            a_lds_block_desc_0,
+            make_tuple(make_pass_through_transform(number<kMPerBlock>{}),
+                       make_merge_transform(make_tuple(number<kKPerBlock>{} / 8, number<8>{}))),
+            make_tuple(sequence<1>{}, sequence<0, 2>{}),
+            make_tuple(sequence<0>{}, sequence<1>{}));
```
In _old ck_ we had really great variable naming convention which made code much, much more easy to understand. Please encode in variable name the tensor descriptor dimensions.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,390 @@
+        return a_lds_block_desc;
+    }
+
+    // 3d + padding
```
This comment seems to be no relevant. Please update.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,390 @@
+
+        // TODO: this 8 is AK1! should be a policy parameter!
+        constexpr auto a_lds_block_desc_0 = make_naive_tensor_descriptor(
+            make_tuple(number<kKPerBlock / 8>{}, number<kMPerBlock>{}, number<8>{}),
```
Shouldn't you use `GetSmemPackA` instead of `8` ? This would be more flexible and would at least adapt to BlockGemm and WarpGemm

ThomasNing replied:
Thank you for the comment Adam! Those are all good suggestions. I have already changed the code based on the comments. Please take a look again.

aosewski reviewed (dismissed):
Great job! Thanks for introducing 2LDS variant. For now I've gone through most of the code, but will yet read more carefully the pipline core part.

ThomasNing committed:
Address the comments

ThomasNing committed:
pre-merge with the develop branch need to fix the bug

illsilin committed:
Merge from internal (#1857)

* enable batched_gemm_softmax_gemm_perm_wmma for gfx12

* disable instances with blocksize=256 in attention examples

* debuggging

* debug

* fixed lds_enabled

* debugging

* Fix and add limit to skiplds feature

* Enable skipLds feature and fix compilation bugs

* add ck_tile definitions for gfx12

* fix clang format and test/wmma_op

* updage instances cmake for gfx12

* disable the test_wmma_op on gfx12

* fix the builds for gfx950

* add gfx12 and gfx950 to default target list

* clean-up cmake file

* Initial introduction of OFP8 data types.

* Renamed FP8 and BF8 tests into FP8_FNUZ and BF8_FNUZ.

* Implementation of ConvertFP32Nearest in test_fp8_ocp.

* Remove dependence on possibly undeclared alias.

* Implement FP8OCP test for stochastic rounding mode.

* Implement FP8OCP tests for half_t type conversions.

* enable bf16 atomic add on gfx950

* Implement ConvertFP32Nearest test.

* Implement ConvertFP32Stochastic test.

* Implement ConvertFP16Nearest and ConvertFP16Stochastic tests.

* Refactoring. Move FP8 definitions into a separate header file.

* Enable easy switching between architectures.

* Fix compilation error for gfx942 architecture.

* Add fp4 type with constants

* only builf gfx950 branch for gfx950 target by default

* Enable OCP build of example_gemm_xdl_fp8.

* Fix formatting.

* fix the build logic for gfx950

* Improve GEMM example verbosity.

* Add constexpr where applicable.

* fix the logic of enabling XDL and WMMA instances

* Improve GEMM example verbosity.

* Enable build of example_gemm_xdl_fp8_bf8 test.

* Fix tests for gfx1101 architecture.

* Build DPP examples only on gfx103 and gfx11 architectures.

* Optionaly run either CPU or GPU verifications with GEMM examples.

* Extend GeneratorTensor_Sequential to produce values of prescribed data types.

* Add missing constructor.

* Add scale type and mxfp conversions

* Update conversions

* Add conversion tests

* Fix typo

* Improve infrastructure for OFP8 data type support.

* BUGFIX. Should not use FP8 as Compute/Accum data type.

* Add custom target for grouped_convnd_bwd_weight tests.

* Can build `tests` target on gfx950.

* Bugfixes on gfx1101 architecture.

* Fix dependencies.

* Add stochastic rounding tests

* Provide single point of truth for FP8 INF and NAN checks

* Prevent instantiation of operators that are not supported by FP8 data types

* Add FP8 type selection into client_axample CMakeLists.txt

* Prevent sccache server from shutting down during build

* Fix test success reporting logic

* Change default verification method to CPU.

GPU verification takes too much time to complete on the emulator.

* Add scale <-> float conversions

* Add scaled conversions with tests

* Add device conversions

* Make sure all tests and examples are built for gfx950

* Facilitate testing of FP8 data types on the emulator

* Introduce two new tensor generators

* Enable instances built for gfx94 to be built on gfx950

* Verify 35_splitk_gemm on floating point numbers.

splitk gemm appears to be losing precision VS reference implementation when FP numbers are involved.

* Format

* Verify 04_gemm_add_add_fastgelu on floating point numbers

* Verify 20_grouped_conv_bwd_weight on floating point numbers

* Verify 38_grouped_conv_bwd_data_multiple_d on floating point numbers

* Verify more tests on floating point data

* Fix data types and improve testing verbocity.

* Add fp4 vectors

* Add debug tests

* Upgrade to NPI 573 build docker.

* Skip on gemm_universal tests.

The tests take too long to complete on the emulator.
Need to see if it is possible to reduce the scope of the testing to just FP8 data types.

* Add new mfma instructions and examples

* Add preprocessor directives for gfx950 specific code

* Fix gfx1101 build

* Document test availability

* Re-enable fp8 gemms for gfx94/95

* Cherry-pick GEMM Universal tests for FP8 data types

* Cleanup

* Add vector types and tests

* Add check_err function

* Add tensor generators

* CK_USE_GFX94 has already been set on this branch

* Fix

* Address formatting issues and leftovers

* Make fail/pass logic consistent within 01_gemm folder

Removed multiple negations in fail/pass logic to propagate `true` as the success indicator.

* Fix GPU verification reporting logic.

* Update year in copyright notice.

* Cleanup

* Use `enum class` instead of `enum`

* Remove set_property for FP8 tests

* Add vector conversions

* Fix

* Fix linker errror

* Clean up

* Fix gfx950 conversions

* Clean up

* Fix more gfx950 conversions

* Fix even more gfx950 conversions

* Narrowing the scope of PR to OCP FP8 enablement only

* Add tests for OCP FP8 vector_type storage

* Fix client examples build

* Fix typo

* Update e8m0 casting

* Rename E8M0 type

* Update unpack method

* Cleanup merge artifacts

* Enable gemm kernel on all gfx9 architectures (#227)

* clean-up

* Implement `non_native_vector_base` with `ext_vector_type` array. (#232)

* Enable support of 1, 2, 4, and 8-byte custom types in CK.

* Fix pool tests for OCP FP8 data type

* Fix build

* Add ckProfiler gemm instances for new mfma instructions and fix ckProfiler build on MI350

* fix clang format

* Add new mfma instructions and examples

* Add preprocessor directives for gfx950 specific code

* Add ckProfiler gemm instances for new mfma instructions and fix ckProfiler build on MI350

* fix clang format

* Fix clang format for the newly merged files

* Use the existing example instances for fp16 bf16 and int8

* Remove comment on new mfma instructions in MfmaInstr

* Update include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* merge from public repo

* Fix ck build

* Fix ck build

* Use double for max_abs_in_val

* Move scaled_type_convert functions to a separate header (#251)

* re-enable building mha lib and gemm_universal_f8 instances for gfx950

* Update library/src/tensor_operation_instance/gpu/CMakeLists.txt

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* fix typo for CK_USE_OCP_FP8

* fix typo for CK_USE_OCP_FP8

* Add FP6 and BF6 types (#261)

* Add a rounding flag

* Add FP6 and BF6

* Add tests

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* Clean up

---------

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* fix one more typo

* Refactor E8M0 scale implementation (#262)

* Refactor E8M0 scale implementation

* Add MXFP6 and MXBF6 conversion methods (#270)

* Add conversions

* Add tests

* Add docstrings

* Add scaled conversions

* Add fp6/bf6 tests

* Remove misleading fp4 test case

* Add docstrings

* Clean up

* Address comments

* Set stricter tolerances for RNE tests

* Add missing tests

* Add native conversions to float

* Revert "Add native conversions to float"

This reverts commit 09467111f73b753c8cc3d597533b187940353dab.

* Update copyright years

* replace the fp6 with bf6 convert calls in test_bf6

* fix test_bf6

* enable smfmac test

* [MX FP8] Add Scaled Type Convert Functions for OCP FP8/BF8 data types (#271)

* Move scaled_type_convert functions to a separate header

* Introduce MX data tests

* Build MX tests only on relevant architectures

* Refactor E8M0 scale implementation

* Fix `config.h` typo

* Cleanup deprecated symbols

* Refactor `amd_ck_fp8.hpp`

* `scaled_type_convert` for `f8_ocp_t`

* Implement test for MX FP8 scaled type convert

* Implement test for MX BF8 scaled type convert

* Scaled type convert for vectors of 2 FP8 elements

* Scaled type convert for vectors of 16 FP8 elements

* Implementation of scaled conversion from F32 to F8

* Add tests for scaled conversions from FP32 to FP8

* Add documentation to the test functions

* Implementation of scaled conversion from F32x2 to F8x2

* Implementation of scaled conversion from F32x16 to F8x16

* Implementation of scaled conversion from F32x32 to F8x32

* Implementation of scaled conversion from F8x32 to F32x32

* Verified on the emulator

* MX FP GEMM - Example Template (#277)

Temporarily uses `DeviceGemmMultiD_ABScale_Xdl_CShuffle_V3` kernel and 128x128 scaling matrices.
Must be modified to use MX-native GEMM kernell with 16 or 32 component vectors per scale.

Verified on the emulator.

* Add vector support

* Add tests

* Add missing type aliases

* Fix test naming

* only build mx example for gfx950

* disable CK_USE_AMD_MFMA_GFX950 by default

* fic build for multiple archs

* fix typo

* fix typo

* Update unpack signature

* Fix merge

* Add size checks in pack function

* Add a flag

* Add conversions

* Fix build logic

* Update pack/unpack methods

* Remove unneeded AsType accessors

* Add docstrings

* Add a flag to config file

* Test the functionality of V_MFMA_F32_16X16X128_F8F6F4 and  V_MFMA_F32_32X32X64_F8F6F4 instructions. (#293)

* Introduced MFMA tests

* Verified f8f6f4 MFMA Instructions

* Move flag logic to scaled_type_convert header

* Use pointers instead of array indices

* Fix a typo

* Update tests and pack functions

* Fix gemm gemm on gfx950

* Fix clang format

* restore the default gput target lists

* fix the jenkinsfile

* add missing ifdef

---------

Co-authored-by: Jing Zhang <jizhan@amd.com>
Co-authored-by: aska-0096 <haocwang@amd.com>
Co-authored-by: Jun Liu <Liu.Jun@amd.com>
Co-authored-by: Andriy Roshchenko <andriy.roshchenko@amd.com>
Co-authored-by: Rostyslav Geyyer <rosty.geyyer@amd.com>
Co-authored-by: Rostyslav Geyyer <46627076+geyyer@users.noreply.github.com>
Co-authored-by: root <root@banff-cyxtera-s83-2.ctr.dcgpu>
Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>
Co-authored-by: jefyang1 <146495389+jefyang1@users.noreply.github.com>
Co-authored-by: jefyang1 <Jeffreyj.Yang@amd.com>

ThomasNing committed:
Finish the integration to develop and have the correct result

ThomasNing committed:
Fix the gtest compilation error

ThomasNing committed:
Fix the gemm_basic error

ThomasNing committed:
clang format

ThomasNing committed:
switch the default pipeline to V3

illsilin committed:
restore cron trigger (#1863)

ThomasNing committed:
fix the benchmark basic script

zjing14 commented:
@ThomasNing Great job!

aosewski commented on example/ck_tile/03_gemm/gemm_basic.hpp:
```diff
@@ -11,21 +11,29 @@
 #define UNIVERSAL_GEMM_PIPELINE ck_tile::BaseGemmPipelineAgBgCrMem
 #define GEMM_PIPELINE_SCHEDULER ck_tile::GemmPipelineScheduler::Interwave
-#elif(CK_TILE_PIPELINE_DEFAULT == CK_TILE_PIPELINE_COMPUTE)
+constexpr bool DoubleSmemBuffer = false;
```
Can we move this into the `*.cpp` file - we have there such config params defined already.

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -455,7 +455,10 @@ struct GemmKernel
+     * @param smem_ptr_0 The start memory pointer of the shared memory block.
      * @param kargs GEMM kernel arguments
+     * @param splitk_batch_offset When there are more than 1 batch needs to split the k.
+     * splitk_batch_offset stands for its the K from which batch.
```
```suggestion
     * @param splitk_batch_offset Utility structure used to calculate k batch offset per workgroup.
```

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -483,15 +486,67 @@ struct GemmKernel
+     * @param smem_ptr_1 The starting pointer of 2nd shared memory block.
+     * @param kargs GEMM kernel arguments
+     * @param splitk_batch_offset When there are more than 1 batch needs to split the k.
+     * splitk_batch_offset stands for its the K from which batch.
```
```suggestion
     * @param splitk_batch_offset Utility structure used to calculate k batch offset per workgroup.
```

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -483,15 +486,67 @@ struct GemmKernel
+
+    /**
+     * @brief Runs single GEMM problem cooperatively by whole workgroup.
+     *
```
Please add `@note` saying that this version is using double LDS & ping-pong algorithm

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -483,15 +486,67 @@ struct GemmKernel
+        // Create Gemm tensor views, pad views and tile windows
+        const auto& gemm_tensor_views_tuple =
+            MakeGemmTensorViews<DstInMemOp>(a_ptr, b_ptr, c_ptr, kargs, splitk_batch_offset);
+        ;
```
Leftover.

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -509,21 +564,50 @@ struct GemmKernel
 
         // allocate LDS
-        __shared__ char smem_ptr[GetSmemSize()];
+        __shared__ char smem_ptr_0[GetSmemSize()];
+        __shared__ char smem_ptr_1[GetSmemSize()];
```
This is not correct. You should rather move this LDS allocation under `RunGemm` and `RunGemm2LDS` functions. We don't want to allocate twice as much LDS in single LDS buffer scenario. Please make sure that in both scenarios there's allocated appropriate amount of memory.

ThomasNing replied:
Unfortunately, we have to do that as our current compiler required us to initialize the shared memory one time before the function call. If we initialize the allocation in the function there will be 3 LDS buffers and will exceed the shared memory (64KB) limitation. Currently, if we have the RunGemm with 64KB LDS allocation, the compiler could still figure out the second shared memory pointer has not been used and successfully have full utilization for the shared memory.

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -509,21 +564,50 @@ struct GemmKernel
         else
         {
-            // Do not compile in case where we have unsupported
-            // VectorSizeC & data type configuration.
```
This comment was added here on purpose - in order to warn before moving this `if constexpr` condition elsewhere, especially into `RunGemm....` functions. We do not want to let to compile `RunGemm...` functions in unsupported scenarios, otherwise there would be compiler errors.

aosewski commented on test/ck_tile/gemm/test_gemm_pipeline.cpp:
```diff
@@ -17,24 +17,25 @@ using Intrawave = ck_tile::integral_constant<ck_tile::GemmPipelineScheduler,
 //                                              ck_tile::GemmPipelineScheduler::Interwave>;
 // using Mem       = ck_tile::integral_constant<GemmPipelineType, GemmPipelineType::Mem>;
-using Comp = ck_tile::integral_constant<GemmPipelineType, GemmPipelineType::Comp>;
+// using CompV4    = ck_tile::integral_constant<GemmPipelineType, GemmPipelineType::CompV4>;
```
Why not enabling v4 ?

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_v1.hpp:
```diff
@@ -39,6 +39,9 @@ struct GemmPipelineAGmemBGmemCRegV1
     static constexpr bool kPadN = Problem::kPadN;
     static constexpr bool kPadK = Problem::kPadK;
 
+    // For the basic gemm pipelien DoubleSmemBuffer set to be false naturally.
+    static constexpr bool DoubleSmemBuffer = false;
+
```
I don't think this is needed here.

ThomasNing replied:
We still need this, because the kernel will read the pipeline configuration to decide if to run the RunGEMM or RunGEMM2LDS.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_mem.hpp:
```diff
@@ -123,6 +123,8 @@ struct GemmPipelineAgBgCrMem : public BaseGemmPipelineAgBgCrMem<Problem>
     static constexpr bool kPadK = Problem::kPadK;
 
+    static constexpr bool isDoubleSmemBuffer = Problem::isDoubleSmemBuffer;
+
```
I don't think this is needed here.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v3.hpp:
```diff
@@ -72,12 +72,12 @@ struct GemmPipelineAgBgCrCompV3 : public BaseGemmPipelineAgBgCrCompV3<Problem>
     static constexpr bool kPadN = Problem::kPadN;
     static constexpr bool kPadK = Problem::kPadK;
 
+    static constexpr bool DoubleSmemBuffer = Problem::DoubleSmemBuffer;
+
```
I don't think this is needed here.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp:
```diff
@@ -19,15 +18,6 @@ struct UniversalGemmPipelineAgBgCrPolicy
     static constexpr auto ATileAccessPattern = tile_distribution_pattern::thread_raked;
     static constexpr auto BTileAccessPattern = tile_distribution_pattern::thread_raked;
 
-    /**
-     * @brief Get the maximum global memory vector load size.
-     *
-     * @tparam Problem      The UniversalGemmPipelineProblem object.
-     * @tparam DataType     The tensor data type we're considering.
-     * @tparam MNPerBlock   The MPerBlock or NPerBlock value depending on tensor (A/B).
-     * @tparam XPerTile     The contiguous Tile dimension size.
-     * @return Maximum DRAM vector load size.
-     */
```
Please do not remove doc.

ThomasNing replied:
Sorry Adam, the copilot automatically delete the comment will add it back.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp:
```diff
@@ -98,18 +88,117 @@ struct UniversalGemmPipelineAgBgCrPolicy
-     *       single thread.
-     *
-     * @return The vector store size for C tensor.
-     */
```
Please do not remove doc.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp:
```diff
@@ -125,8 +214,6 @@ struct UniversalGemmPipelineAgBgCrPolicy
             if constexpr(TransposeC)
             {
-                // In this case each thread has multiple consecutive elements in
-                // N dimension, however consecutive threads' elements have stride.
```
Please do not remove doc!

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp:
```diff
@@ -136,7 +223,6 @@ struct UniversalGemmPipelineAgBgCrPolicy
             }
             else
             {
-                // In this case each thread has just a single item in Ndim
```
Please do not remove doc!

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp:
```diff
@@ -145,13 +231,10 @@ struct UniversalGemmPipelineAgBgCrPolicy
         {
             if constexpr(TransposeC)
             {
-                // In this case each thread has just a single item in Mdim
```
Please do not remove doc!

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp:
```diff
@@ -145,13 +231,10 @@ struct UniversalGemmPipelineAgBgCrPolicy
             else
             {
-                // In this case each thread has multiple consecutive elements in
-                // M dimension, however consecutive threads' elements have stride.
```
Please do not remove doc!

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_universal_pipeline_ag_bg_cr_policy.hpp:
```diff
@@ -425,16 +508,20 @@ struct UniversalGemmPipelineAgBgCrPolicy
+        constexpr index_t smem_size_a =
+            integer_least_multiple(sizeof(typename Problem::ADataType) *
+                                       MakeALdsBlockDescriptor<Problem>().get_element_space_size(),
+                                   16);
```
Can we make from this value a policy member and reuse everywhere ?

ThomasNing replied:
Not now, because not every pipeline is using the same ALdsBlockDescriptor. We could modify here after they are all unified.

aosewski commented on test/ck_tile/gemm/test_gemm_pipeline_util.hpp:
```diff
@@ -69,16 +72,24 @@ class TestCkTileGemmPipeline : public ::testing::Test
+            ck_tile::BaseGemmPipelineAgBgCrMem<GemmPipelineProblem>,
+            std::conditional_t<PipelineType == GemmPipelineType::CompV3,
+                               ck_tile::BaseGemmPipelineAgBgCrCompV3<GemmPipelineProblem>,
+                               ck_tile::BaseGemmPipelineAgBgCrCompV4<GemmPipelineProblem>>>;
```
Probably soon we would have even more pipelines. Then this approach wouldn't be maintainable. Please think about ie. using a map with keys as enumerators & values as pipeline types, or other more flexible approach.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,186 @@
+// UniversalGemm Pipeline Policy.
+// Default policy class should not be templated, put template on
+// member functions instead.
+struct GemmPipelineAGmemBGmemCregComputeV4DefaultPolicy : public UniversalGemmBasePolicy
```
Could we actually be more succint with naming? Why not to continue universal gemm notation ? `g` stands for global, `s` stands for LDS and `r` stands for registers. It would be much shorter and still verbose. The same for file name.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,186 @@
+struct GemmPipelineAGmemBGmemCregComputeV4DefaultPolicy : public UniversalGemmBasePolicy
+{
+    template <typename Problem>
+    CK_TILE_HOST_DEVICE static constexpr auto GetVectorSizeC()
```
Why this function is not actually in UniversalGemmBasePolicy ? It's the same for other universal gemm pipelines for now..

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,186 @@
+        using BlockGemm         = decltype(GetBlockGemm<Problem>());
+        constexpr index_t KPack = BlockGemm::KPack;
+        return KPack;
+    }
```
As above.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_compute_v4_policy.hpp:
```diff
@@ -0,0 +1,186 @@
+        constexpr index_t smem_size_a =
+            integer_least_multiple(sizeof(typename Problem::ADataType) *
+                                       MakeALdsBlockDescriptor<Problem>().get_element_space_size(),
+                                   16);
```
You could reuse here Policy alignment member or function for that.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v4.hpp:
```diff
@@ -0,0 +1,555 @@
+    }
+};
+
+// Compute optimized pipeline version 4
+// The difference between this pipeline and compute version 3 is it has two LDS window that will use
+// the ping-pong buffer to grab memory from the global memory. While one LDS is grabbing the data
+// from global memory, the other will call the warps on running the MFMA matrix multiplication. When
+// the matrix is in bigger shape, it will keep the Warp always busy and cover the memory loading
+// time. It will have better performance comparing to the Compute Version 3 when they have the same
+// block tile and better performance when you have M, N, K all > 8K even when the compute V3 block
+// size is 2 times of the compute V4.
```
Could you format this using doxygen doc style ?

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v4.hpp:
```diff
@@ -0,0 +1,555 @@
+template <typename Problem, typename Policy = GemmPipelineAGmemBGmemCregComputeV4DefaultPolicy>
+struct GemmPipelineAgBgCrCompV4 : public BaseGemmPipelineAgBgCrCompV4<Problem>
+{
+    using Base             = BaseGemmPipelineAgBgCrCompV3<Problem>;
```
Spelling! You'd rather meant to use V4
```suggestion
    using Base             = BaseGemmPipelineAgBgCrCompV4<Problem>;
```

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v4.hpp:
```diff
@@ -0,0 +1,555 @@
+                                   const BDramBlockWindowTmp& b_dram_block_window_tmp,
+                                   const BElementFunction& b_element_func,
+                                   index_t num_loop,
+                                   void* p_smem_0,
+                                   void* p_smem_1) const
```
don't you need `__restrict__` here aswell ?

jakpiase committed:
add vectorloads on non-k dim for memory pipelines (#1856)

ThomasNing committed:
Solving the Review comments

kylasa committed:
Support for dtypes (fp8, bf8, bf16 and fp16) for the ck_tile/03_gemm example. (#1845)

* Support bf16/fb8/bf8 datatypes for ck_tile/gemm

* remove commented out code.

* Addressing code review comments and enabling universal_gemm for all the supported data types.

* Merge conflict resolution.

* Solve the memory pipeline compilation error. Merge with the new change of CShuffle

* finish the feature, pass the tests

* Fix the pipeline and add the benchmark script for other data types

---------

Co-authored-by: ThomasNing <thomas.ning@amd.com>

aosewski committed:
CK Tile - small fix to hotloop scheduler & KPack value. (#1867)

* Use SmemPack in HotLoop scheduler

* Additional debug print information

* Change KPack value.

Hardcode for now, as without AK1/BK1 there's no good way to determine
its value.

* Fix HotLoopScheduler MFMA instr parameters.

ThomasNing committed:
address the new comments

ThomasNing committed:
fix a small bug on the old

ThomasNing commented:
Previous comments resolved.

geyyer committed:
Add a host mx gemm reference kernel (#1864)

* Add mx gemm reference kernel

* Update copyright year

* Update mx gemm example

* Use element-wise ops in the reference gemm

danielsu-amd committed:
External CI: enable amd-develop branch trigger (#1859)

ThomasNing committed:
merge with the develop support the fp8 with computev4

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

ThomasNing committed:
Merge branch 'develop' into ck_tile/gemm_compute_v4

ThomasNing committed:
Solve FMHA error

ThomasNing committed:
clang format

ThomasNing committed:
Fix the memory pipleine

illsilin committed:
Merge branch 'develop' into ck_tile/gemm_compute_v4

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -457,8 +457,8 @@ struct GemmKernel
      * @param kargs GEMM kernel arguments
-     * @param splitk_batch_offset When there are more than 1 batch needs to split the k.
-     * splitk_batch_offset stands for its the K from which batch.
+     * @param splitk_batch_offset param splitk_batch_offset Utility structure used to calculate k
```
```suggestion
     * @param splitk_batch_offset Utility structure used to calculate k
```

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -501,14 +501,16 @@ struct GemmKernel
      * @param kargs GEMM kernel arguments
-     * @param splitk_batch_offset When there are more than 1 batch needs to split the k.
-     * splitk_batch_offset stands for its the K from which batch.
+     * @param splitk_batch_offset splitk_batch_offset Utility structure used to calculate k batch
```
```suggestion
     * @param splitk_batch_offset Utility structure used to calculate k batch
```

aosewski commented on test/ck_tile/gemm/test_gemm_pipeline_util.hpp:
```diff
@@ -84,12 +108,8 @@ class TestCkTileGemmPipeline : public ::testing::Test
-                               ck_tile::BaseGemmPipelineAgBgCrCompV3<GemmPipelineProblem>,
-                               ck_tile::BaseGemmPipelineAgBgCrCompV4<GemmPipelineProblem>>>;
+        using BaseGemmPipeline =
+            typename GemmPipelineTypeSelector<PipelineType, GemmPipelineProblem>::base_pipeline;
```
Great solution!

aosewski commented on test/ck_tile/gemm/test_gemm_pipeline.cpp:
```diff
@@ -14,29 +14,26 @@ using Row       = ck_tile::tensor_layout::gemm::RowMajor;
+    std::tuple<    Col,     Row,     Row,       F16,       F16,         F32,       F16,             Interwave,         Mem>,
+    std::tuple<    Col,     Col,     Row,       F16,       F16,         F32,       F16,             Intrawave,         Mem>,
+    std::tuple<    Col,     Col,     Row,       F16,       F16,         F32,       F16,             Intrawave,        CompV3>,
+    std::tuple<    Col,     Col,     Row,       F16,       F16,         F32,       F16,             Interwave,         Mem>
```
Please add V4 pipeline here as well.

spolifroni-amd reviewed (commented):
Nothing to review for docs.

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

ThomasNing committed:
Solve merge conflict and add the gtest for compv4

ThomasNing committed:
Merge branch 'develop' of https://github.com/ROCm/composable_kernel into develop

ThomasNing committed:
Merge branch 'develop' into ck_tile/gemm_compute_v4

ThomasNing committed:
sync with develop
